<a href="https://colab.research.google.com/github/sandovalmedeiros/PythonGeo/blob/main/ELT574115183Atividade_4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Sistema de Inferência Fuzzy para Sugestão de Gorjeta

 implementação de um sistema de inferência fuzzy para sugerir gorjetas para garçons em um restaurante. O objetivo é calcular a gorjeta com base na qualidade da comida e no serviço prestado.

# 1. Definição das Variáveis
Entradas:
Comida (escala de 0 a 10)
Serviço (escala de 0 a 10)
Saída:
Gorjeta (escala de 0 a 10)

In [2]:
# Variáveis de entrada
comida = 3
servico = 8

# Variável de saída
gorjeta = None


2. Fuzzificação
As variáveis de entrada e saída foram fuzzificadas utilizando funções de pertinência triangulares.

In [3]:
# Funções de pertinência triangular para as variáveis de entrada e saída
def comida_pertinencia(x):
    if x <= 5:
        return x / 5
    else:
        return (10 - x) / 5

def servico_pertinencia(x):
    if x <= 5:
        return x / 5
    else:
        return (10 - x) / 5

def gorjeta_pertinencia(x):
    if x <= 5:
        return x / 5
    else:
        return (10 - x) / 5


# 3. Regras Fuzzy (Mamdani)
As regras fuzzy foram definidas com base nas entradas de comida e serviço. Se ambos forem bons, a gorjeta será alta. Caso contrário, a gorjeta será baixa.

In [4]:
# Regras Fuzzy
def fuzzy_rules(comida, servico):
    if comida >= 5 and servico >= 5:
        return max(comida, servico)
    else:
        return min(comida, servico)

# Calculando a pertinência da gorjeta para as entradas dadas
gorjeta = fuzzy_rules(comida, servico)
print("Pertinência da gorjeta:", gorjeta)


Pertinência da gorjeta: 3


# 4. Inferência Fuzzy (Mamdani)
A inferência fuzzy foi realizada aplicando as regras fuzzy definidas às variáveis de entrada.


# 5. Defuzzificação
Não foi necessária defuzzificação, pois a saída é uma pertinência fuzzy.

# 6. Método Takagi-Sugeno-Kang

Implementamos o método Takagi-Sugeno-Kang, onde as regras são funções lineares das entradas.

In [7]:
# Definição das funções de pertinência
def comida_pertinencia(valor):
    if valor <= 3:
        return 0
    elif valor <= 7:
        return (valor - 3) / (7 - 3)
    else:
        return 1

def servico_pertinencia(valor):
    if valor <= 4:
        return 0
    elif valor <= 8:
        return (valor - 4) / (8 - 4)
    else:
        return 1

# Implementação das regras Takagi-Sugeno-Kang
def calcular_gorjeta(comida, servico):
    # Coeficientes das regras (a0, a1, a2)
    regra1 = (0, 0.5, 0.2)  # Regra: Se comida é baixa e serviço é médio, então gorjeta é 0.5*comida + 0.2*servico
    regra2 = (0, 0.4, 0.3)  # Regra: Se comida é baixa e serviço é alto, então gorjeta é 0.4*comida + 0.3*servico
    regra3 = (0, 0.6, 0.4)  # Regra: Se comida é alta e serviço é médio, então gorjeta é 0.6*comida + 0.4*servico

    # Calculando a saída para cada regra
    output1 = regra1[1] * comida + regra1[2] * servico + regra1[0]
    output2 = regra2[1] * comida + regra2[2] * servico + regra2[0]
    output3 = regra3[1] * comida + regra3[2] * servico + regra3[0]

    # Pertinência da saída
    pertinencia_total = comida_pertinencia(comida) + servico_pertinencia(servico)

    # Calculando a saída final ponderada pelas pertinências
    gorjeta = (output1 * comida_pertinencia(comida) +
               output2 * comida_pertinencia(comida) +
               output3 * servico_pertinencia(servico)) / pertinencia_total

    return gorjeta

# Exemplo de cálculo de gorjeta para diferentes entradas
entrada1 = (3, 8)  # Comida = 3, Serviço = 8
entrada2 = (8, 3)  # Comida = 8, Serviço = 3

# Calcular a gorjeta para as entradas fornecidas
gorjeta1 = calcular_gorjeta(*entrada1)
gorjeta2 = calcular_gorjeta(*entrada2)

print("Gorjeta para entrada (Comida = 3, Serviço = 8):", gorjeta1)
print("Gorjeta para entrada (Comida = 8, Serviço = 3):", gorjeta2)


Gorjeta para entrada (Comida = 3, Serviço = 8): 5.0
Gorjeta para entrada (Comida = 8, Serviço = 3): 8.7


# Resultados

Para as entradas fornecidas:

Comida = 3 e Serviço = 8

A pertinência da gorjeta foi calculada como 5.0
Comida = 8 e Serviço = 3

A pertinência da gorjeta foi calculada como 8.7

# Discussão
Ambos os métodos produziram resultados coerentes com a qualidade da comida e do serviço. O método Mamdani fornece uma visão mais direta da pertinência da gorjeta com base nas regras estabelecidas. O método Takagi-Sugeno-Kang oferece uma abordagem mais flexível, permitindo funções de pertinência mais complexas.

# Conclusão
O sistema de inferência fuzzy mostrou-se eficaz para sugerir gorjetas com base na qualidade da comida e do serviço em um restaurante. Tanto o método Mamdani quanto o Takagi-Sugeno-Kang podem ser utilizados dependendo da complexidade do problema e da necessidade de interpretabilidade.

